File: examples/java_modern_module/ModuleMain.java
package com.example.module;

import android.content.SharedPreferences;
import android.os.Build;
import android.os.Bundle;
import android.os.ParcelFileDescriptor;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.lang.reflect.Constructor;
import java.lang.reflect.Method;
import java.util.stream.Collectors;

import io.github.libxposed.api.XposedInterface;
import io.github.libxposed.api.XposedModule;

/**
 * Modern LibXposed module entry in Java.
 * Declared in: src/main/resources/META-INF/xposed/java_init.list
 */
public class ModuleMain extends XposedModule {
    private static final String TAG = "JavaXposedModule";

    @Override
    public void onModuleLoaded(@NonNull ModuleLoadedParam param) {
        log(Log.INFO, TAG, "Module loaded in: " + param.getProcessName());
        log(Log.INFO, TAG, "Framework: " + getFrameworkName() + " (API " + getApiVersion() + ")");
    }

    @Override
    @RequiresApi(Build.VERSION_CODES.Q)
    public void onPackageLoaded(@NonNull PackageLoadedParam param) {
        log(Log.INFO, TAG, "Package loaded: " + param.getPackageName());
    }

    @Override
    public void onPackageReady(@NonNull PackageReadyParam param) {
        if (!param.isFirstPackage()) return;

        ClassLoader classLoader = param.getClassLoader();

        // 1. Remote Preferences & Live Listener
        SharedPreferences prefs = getRemotePreferences("main_settings");
        boolean isEnabled = prefs.getBoolean("enable_feature", false);
        log(Log.INFO, TAG, "enable_feature is: " + isEnabled);

        prefs.registerOnSharedPreferenceChangeListener((sp, key) -> {
            log(Log.INFO, TAG, "Config updated for key: " + key + " -> " + sp.getBoolean(key, false));
        });

        // 2. Remote Files
        try (ParcelFileDescriptor pfd = openRemoteFile("config.json")) {
            try (BufferedReader reader = new BufferedReader(new FileReader(pfd.getFileDescriptor()))) {
                String content = reader.lines().collect(Collectors.joining("\n"));
                log(Log.INFO, TAG, "Read remote config.json:\n" + content);
            }
        } catch (FileNotFoundException e) {
            log(Log.WARN, TAG, "config.json not found");
        } catch (Exception e) {
            log(Log.ERROR, TAG, "Error reading remote file", e);
        }

        // 3. Interceptor Chain Hooks & Invokers
        try {
            Class<?> targetClass = Class.forName("com.target.app.SecurityService", true, classLoader);
            Method isRootedDevice = targetClass.getDeclaredMethod("isRootedDevice");
            Constructor<?> ctor = targetClass.getDeclaredConstructor();

            // Hook method
            hook(isRootedDevice)
                    .setId("root_check_hook")
                    .setPriority(XposedInterface.PRIORITY_DEFAULT)
                    .setExceptionMode(XposedInterface.ExceptionMode.PROTECTIVE)
                    .intercept(chain -> {
                        log(Log.INFO, TAG, "Intercepted isRootedDevice check!");
                        // Bypass original check and return false
                        return false;
                    });

            // Hook Constructor
            hook(ctor)
                    .setPriority(XposedInterface.PRIORITY_HIGHEST)
                    .setExceptionMode(XposedInterface.ExceptionMode.PASSTHROUGH)
                    .intercept(chain -> {
                        log(Log.INFO, TAG, "SecurityService instance initializing");
                        chain.proceed();
                        return null; // For void methods and constructors, return null
                    });

            // Invoker demonstration: call raw method bypassing hooks
            Object rawResult = getInvoker(isRootedDevice)
                    .setType(XposedInterface.Invoker.Type.ORIGIN)
                    .invoke(null);

        } catch (ReflectiveOperationException e) {
            log(Log.ERROR, TAG, "Failed to resolve target methods", e);
        }
    }

    @Override
    public boolean onHotReloading(@NonNull HotReloadingParam param) {
        log(Log.INFO, TAG, "onHotReloading: saving state...");
        Bundle state = new Bundle();
        state.putLong("reload_timestamp", System.currentTimeMillis());
        param.setSavedInstanceState(state);
        return true;
    }

    @Override
    public void onHotReloaded(@NonNull HotReloadedParam param) {
        log(Log.INFO, TAG, "onHotReloaded: restoring state in " + param.getProcessName());

        // Replace old hooks atomically
        for (HookHandle handle : param.getOldHookHandles()) {
            if ("root_check_hook".equals(handle.getId())) {
                handle.replaceHook(chain -> {
                    log(Log.INFO, TAG, "Hot-reloaded root check interceptor!");
                    return false;
                });
            } else {
                handle.unhook();
            }
        }
    }
}

File: examples/kotlin_modern_module/ModuleMain.kt
package com.example.module

import android.os.Build
import android.os.Bundle
import android.util.Log
import androidx.annotation.RequiresApi
import io.github.libxposed.api.XposedInterface.ExceptionMode
import io.github.libxposed.api.XposedInterface.Invoker
import io.github.libxposed.api.XposedModule
import io.github.libxposed.api.XposedModuleInterface.HotReloadedParam
import io.github.libxposed.api.XposedModuleInterface.HotReloadingParam
import io.github.libxposed.api.XposedModuleInterface.ModuleLoadedParam
import io.github.libxposed.api.XposedModuleInterface.PackageLoadedParam
import io.github.libxposed.api.XposedModuleInterface.PackageReadyParam
import io.github.libxposed.api.XposedModuleInterface.SystemServerStartingParam
import java.io.FileNotFoundException
import java.io.FileReader

/**
 * Modern LibXposed module entry in Kotlin.
 * Declared in: src/main/resources/META-INF/xposed/java_init.list
 */
class ModuleMain : XposedModule() {

    companion object {
        const val TAG = "ModernXposedModule"
    }

    // =========================================================================
    // 1. Lifecycle: Module Loaded into Process
    // =========================================================================
    override fun onModuleLoaded(param: ModuleLoadedParam) {
        log(Log.INFO, TAG, "Module loaded into process: ${param.processName}")
        log(Log.INFO, TAG, "Framework: $frameworkName v$frameworkVersion (Code: $frameworkVersionCode, API: $apiVersion)")

        val props = frameworkProperties
        val hasProp: (Long) -> Boolean = { (props and it) != 0L }
        log(Log.INFO, TAG, "Capabilities -> System: ${hasProp(PROP_CAP_SYSTEM)}, Remote: ${hasProp(PROP_CAP_REMOTE)}, API Protection: ${hasProp(PROP_RT_API_PROTECTION)}")
    }

    // =========================================================================
    // 2. Lifecycle: Android Q+ Default ClassLoader Available
    // =========================================================================
    @RequiresApi(Build.VERSION_CODES.Q)
    override fun onPackageLoaded(param: PackageLoadedParam) {
        log(Log.INFO, TAG, "Package loaded: ${param.packageName}, firstPackage: ${param.isFirstPackage}")
    }

    // =========================================================================
    // 3. Lifecycle: System Server (Hooking OS Framework Services)
    // =========================================================================
    override fun onSystemServerStarting(param: SystemServerStartingParam) {
        log(Log.INFO, TAG, "System server starting. System ClassLoader: ${param.classLoader}")
    }

    // =========================================================================
    // 4. Primary Hook Entry Point: Application ClassLoader Ready
    // =========================================================================
    override fun onPackageReady(param: PackageReadyParam) {
        log(Log.INFO, TAG, "Package ready to run: ${param.packageName}")
        if (!param.isFirstPackage) return

        val appClassLoader = param.classLoader

        // ---------------------------------------------------------------------
        // Remote Preferences (Read-Only in Target) & Live Change Listener
        // ---------------------------------------------------------------------
        val prefs = getRemotePreferences("main_settings")
        val isFeatureActive = prefs.getBoolean("enable_feature", true)
        log(Log.INFO, TAG, "Initial config 'enable_feature': $isFeatureActive")

        prefs.registerOnSharedPreferenceChangeListener { sharedPreferences, key ->
            val updatedVal = sharedPreferences.getBoolean(key, false)
            log(Log.INFO, TAG, "Real-time config update -> $key: $updatedVal")
        }

        // ---------------------------------------------------------------------
        // Remote Files (Binary Blobs & Large Data)
        // ---------------------------------------------------------------------
        try {
            val content = openRemoteFile("custom_rules.json").use { pfd ->
                FileReader(pfd.fileDescriptor).readText()
            }
            log(Log.INFO, TAG, "Loaded remote rules:\n$content")
        } catch (_: FileNotFoundException) {
            log(Log.INFO, TAG, "Remote file custom_rules.json not found")
        }

        // ---------------------------------------------------------------------
        // Interceptor Chain Hooking
        // ---------------------------------------------------------------------
        try {
            val targetClass = Class.forName("com.target.app.AuthManager", true, appClassLoader)
            val authMethod = targetClass.getDeclaredMethod("authenticate", String::class.java, String::class.java)
            val authConstructor = targetClass.getDeclaredConstructor()

            // Hook Method with Interceptor Chain
            hook(authMethod)
                .setId("auth_method_hook")
                .setPriority(PRIORITY_DEFAULT)
                .setExceptionMode(ExceptionMode.PROTECTIVE)
                .intercept { chain ->
                    log(Log.INFO, TAG, "Auth called with user: ${chain.getArg(0)}")

                    // 1. Proceed with original arguments
                    val originalResult = chain.proceed() as Boolean

                    // 2. Or proceed with modified arguments:
                    // val newArgs = arrayOf("spoofed_user", chain.getArg(1))
                    // val result = chain.proceed(newArgs) as Boolean

                    // 3. Or invoke raw method directly (bypassing all hooks) via Invoker:
                    val rawResult = getInvoker(authMethod)
                        .setType(Invoker.Type.ORIGIN)
                        .invoke(chain.thisObject, "admin", "token")

                    // Return spoofed/modified boolean
                    true
                }

            // Hook Constructor
            hook(authConstructor)
                .setPriority(PRIORITY_HIGHEST)
                .setExceptionMode(ExceptionMode.PASSTHROUGH)
                .intercept { chain ->
                    log(Log.INFO, TAG, "Instantiating AuthManager...")
                    chain.proceed()
                    // Return null/Unit for constructors
                }

        } catch (t: Throwable) {
            log(Log.ERROR, TAG, "Failed to initialize hooks in onPackageReady", t)
        }
    }

    // =========================================================================
    // 5. Hot Reloading: Old Generation Retirement (Runs in Old Bytecode)
    // =========================================================================
    override fun onHotReloading(param: HotReloadingParam): Boolean {
        log(Log.INFO, TAG, "Preparing old module generation for hot reload...")

        // Save classloader-neutral state for the incoming new generation
        val state = Bundle().apply {
            putLong("hot_reload_time", System.currentTimeMillis())
        }
        param.setSavedInstanceState(state)

        // Return true to allow hot reload
        return true
    }

    // =========================================================================
    // 6. Hot Reloading: New Generation Activation (Runs in New Bytecode)
    // =========================================================================
    override fun onHotReloaded(param: HotReloadedParam) {
        log(Log.INFO, TAG, "New module generation activated in ${param.processName}!")

        val state = param.savedInstanceState as? Bundle
        val lastReloadTime = state?.getLong("hot_reload_time") ?: 0L
        log(Log.INFO, TAG, "State restored. Previous reload timestamp: $lastReloadTime")

        // Atomically replace old hooks
        for (handle in param.oldHookHandles) {
            if ("auth_method_hook" == handle.id) {
                handle.replaceHook { chain ->
                    log(Log.INFO, TAG, "Updated hot-reloaded auth interceptor executing!")
                    chain.proceed()
                }
            } else {
                handle.unhook()
            }
        }
    }
}
